lib/archethic/account/mem_tables/token_ledger.ex

```diff
@@ -130,34 +135,59 @@ defmodule Archethic.Account.MemTables.TokenLedger do
   """
   @spec spend_all_unspent_outputs(binary()) :: :ok
   def spend_all_unspent_outputs(address) do
+    {:ok, pid} = DB.start_inputs_writer(:token, address)
+
     @unspent_output_index_table
     |> :ets.lookup(address)
-    |> Enum.each(fn {_, from, token_address, token_id} ->
-      :ets.update_element(@ledger_table, {address, from, token_address, token_id}, {3, true})
+    |> Enum.each(fn {to, from, token_address, token_id} ->
+      [{_, amount, _, timestamp, protocol_version}] =
+        :ets.lookup(@ledger_table, {to, from, token_address, token_id})
+
+      DB.append_input(pid, %VersionedTransactionInput{
+        protocol_version: protocol_version,
+        input: %TransactionInput{
+          from: from,
+          amount: amount,
+          spent?: true,
+          timestamp: timestamp,
+          type: {:token, token_address, token_id}
+        }
+      })
+
+      :ets.delete(@ledger_table, {address, from, token_address, token_id})
     end)
+
+    :ets.delete(@unspent_output_index_table, address)
+
+    DB.stop_inputs_writer(pid)
   end
 
   @doc """
   Retrieve the entire inputs for a given address (spent or unspent)
   """
   @spec get_inputs(binary()) :: list(VersionedTransactionInput.t())
   def get_inputs(address) when is_binary(address) do
+    spent = DB.get_inputs(:token, address)
```

Comment:
I think it's not needed to always read the DB and the ETS table as the inputs will be either in the ETS table, either in the DB but never both in the same time for an address.
So you may first read the ETS table to retrieve the inputs, and if there is nothing in then you can go throught DB.

Reply:
I think on the explorer it's just a view, but when there is a new transaction, in `spend_all_unspent_outputs` you delete all unspent outputs from the ETS table and write them in DB.
In the explorer, we check if an output is spent or not based on the last unspent output of the chain. But in the ETS table there is only the last transaction unspent ouputs I think.

Reply:
You are correct! and that's why I had to do the #627. I finally understand fully.